src/tls_relation.py

```diff
@@ -138,70 +137,57 @@ def get_hostname_from_csr(self, tls_relation: Relation, app: Application, csr: s
                 return hostname
         return ""
 
-    def get_tls_relation(self, charm: CharmBase) -> Union[Relation, None]:
+    def get_tls_relation(self) -> Union[Relation, None]:
         """Get the TLS certificates relation.
 
-        Args:
-            charm: The Juju charm containing the relation
-
         Returns:
             The TLS certificates relation of the charm.
         """
-        relation = charm.model.get_relation(TLS_CERT)
+        relation = self.charm_model.get_relation(TLS_CERT)
         return relation
 
     # The charm will not have annotations to avoid circular imports.
     def certificate_relation_joined(  # type: ignore[no-untyped-def]
-        self, hostname: str, charm
+        self,
+        hostname: str,
+        certificates: TLSCertificatesRequiresV2,
     ) -> None:
         """Handle the TLS Certificate joined event.
 
         Args:
             hostname: Certificate's hostname.
-            charm: The Ingress charm that has the TLS relation
+            certificates: The certificate requirer library instance.
         """
-        tls_certificates_relation = self.get_tls_relation(charm)
-        peer_relation = charm.model.get_relation(PEER_RELATION_NAME)
+        tls_certificates_relation = self.get_tls_relation()
+        peer_relation = self.charm_model.get_relation(PEER_RELATION_NAME)
         assert isinstance(tls_certificates_relation, Relation)  # nosec
-        private_key_dict = {}
-        if JujuVersion.from_environ().has_secrets:
-            secret = charm.model.get_secret(label=f"private-key-{hostname}")
-            secret.grant(tls_certificates_relation)
-            private_key_dict["key"] = secret.get_content()["key"].encode()
-            private_key_dict["password"] = secret.get_content()["password"].encode()
-        else:
-            private_key_dict["key"] = self.get_relation_data_field(
-                f"key-{hostname}", peer_relation, charm.app
-            ).encode()
-            private_key_dict["password"] = self.get_relation_data_field(
-                f"password-{hostname}", peer_relation, charm.app
-            ).encode()
+        assert isinstance(peer_relation, Relation)  # nosec
+        private_key_dict = self._get_private_key(hostname)
         csr = generate_csr(
-            private_key=private_key_dict["key"],
-            private_key_password=private_key_dict["password"],
+            private_key=private_key_dict["key"].encode(),
+            private_key_password=private_key_dict["password"].encode(),
             subject=hostname,
+            sans_dns=[hostname],
```

Comment:
Cool, it would be interesting to have a config for defining SANs in certificate.

Reply:
Well not sure that would be relevant, we took the decision to issue 1 certificate per domain managed by nginx-ingress-integrator.
The main reasoning is that let's encrypt has pretty low rate limits in terms of certificate revocation, and when a domain is added or removed, if we have all the domains in the same certificate it requires to revoke it.
On the other hands 1 certificate per domain requires more certificates, but the rate limit is higher on let's encrypt.

The SAN here was added because recent versions of Nginx Ingress Controllers complains that Subject is deprecated and that we should use SAN